core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -51,6 +51,10 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
   private var _stack          = null.asInstanceOf[Array[Continuation]]
   private var _stackSize      = 0
   private val emptyTrace      = Trace.empty
+  private val nForkedFibers   = new AtomicInteger(0)
```

Comment:
> Also, I am not sure why you need atomic integer here

Changed this to `var` 👍 

> then in theory, there is a design that permits you to store this locally, rather than on the heap, similar to opCount and the like in the run loop

I tried a few different things here, and the options as I see them are:

1. We add a new case class `Fork` which extends `ZIO`. This is very similar to what `Sync` is, but allows us to keep track of how many fibers we've forked within the `FiberRuntime` runloop.
2. We add a new param to `Sync` to identify whether it contains a forked fiber - we probably don't want to do that though
3. We have a counter and externally increment it - this is what we currently do

I'm not going to discuss (2) as I think having another argument on Sync is going to add too much overhead.

From how I see it, (1) is a decent option, but the part I'm struggling with is how to propagate the information on how many fibers have been forked downstream once we need to call `runLoop` within the runloop itself. One main issue with this is that we normally call `.fork` as part of `WhileLoop`, which calls `runLoop` for each element instead of relying on the current depth loop. We could return that as part of the method's response, but that adds extra allocations - so we still have to use the heap to store that info 😕 Or maybe refactor how WhileLoop works? 

The other thing, having 1 more extension to ZIO will likely slow down the runloop a bit for cases that there aren't any forked fibers, since the pattern-matching will contain an extra clause. Not sure how big of a deal that might be though.

Now to (3). I think this is the option that has the least overhead for cases that we're not forking fibers often.  However, something makes me kind of worried is that I don't think the logic to determine whether we should yield or not should leak outside of the `FiberRuntime` itself, so perhaps (1) is a better option on that front.